chiselFrontend/src/main/scala/chisel3/internal/firrtl/IR.scala

```diff
@@ -257,6 +229,276 @@ object MemPortDirection {
+  }
+
+  private def doFirrtlOp(op: firrtlir.PrimOp, that: IntervalRange): IntervalRange = {
+    PrimOps.set_primop_type(
+      firrtlir.DoPrim(op,
+        Seq(firrtlir.Reference("a", this), firrtlir.Reference("b", that)), Nil,firrtlir.UnknownType)
+    ).tpe match {
+      case i: firrtlir.IntervalType => IntervalRange(i.lower, i.upper, i.point)
+      case other => sys.error("BAD!")
+    }
+  }
+
+  private def doFirrtlOp(op: firrtlir.PrimOp, that: Int): IntervalRange = {
+    PrimOps.set_primop_type(
+      firrtlir.DoPrim(op,
+        Seq(firrtlir.Reference("a", this)), Seq(BigInt(that)), firrtlir.UnknownType)
+    ).tpe match {
+      case i: firrtlir.IntervalType => IntervalRange(i.lower, i.upper, i.point)
+      case other => sys.error("BAD!")
+    }
+  }
+
+  override def +&(that: IntervalRange): IntervalRange = {
+    //    doFirrtlOp(PrimOps.Add, that)
+    IntervalRange(firrtlir.UnknownBound, firrtlir.UnknownBound, firrtlir.UnknownWidth)
+  }
+  private def getRange(op: firrtlir.PrimOp, tpe1: IntervalRange, tpe2: IntervalRange): IntervalRange = ???
+  override def -&(that: IntervalRange): IntervalRange = {
+    //    doFirrtlOp(PrimOps.Sub, that)
+    IntervalRange(firrtlir.UnknownBound, firrtlir.UnknownBound, firrtlir.UnknownWidth)
+  }
+
+  private def shiftLeft(bound: firrtlir.Bound, n: Int): firrtlir.Bound = {
+    if(n < 1 ) {
+      bound
+    }
+    else {
+      val multiplier = BigDecimal(BigInt(1) << n)
+      bound match {
+        case firrtlir.Open(x) => firrtlir.Open(x * multiplier)
+        case firrtlir.Closed(x) => firrtlir.Closed(x * multiplier)
+        case _ => firrtlir.UnknownBound
+      }
+    }
+  }
+
+  private def shiftRight(bound: firrtlir.Bound, n: Int, binaryPoint: BinaryPoint): firrtlir.Bound = {
+    if(n < 1 ) {
+      bound
+    }
+    else {
+      // TODO intervals chick -- can add bp info (if known); see firrtl primops
+      // Because of loss of significant digits, range changes too BUT to figure that out, you need to know BP
+      binaryPoint match {
+        case UnknownBinaryPoint => firrtlir.UnknownBound
+        case KnownBinaryPoint(value) =>
+          val divisor = 1 << n
+          bound match {
+            case firrtlir.Open(x)   => firrtlir.Open(x / divisor)
+            case firrtlir.Closed(x) => firrtlir.Closed(x / divisor)
+            case _ => bound
+          }
+      }
+
+    }
+  }
+
+  override def <<(that: Int): IntervalRange = {
+    IntervalRange(
+      shiftLeft(lowerBound, that),
+      shiftLeft(upperBound, that),
+      binaryPoint
+    )
+  }
+
+  override def >>(that: Int): IntervalRange = {
+    IntervalRange(
+      shiftRight(lowerBound, that, binaryPoint),
+      shiftRight(upperBound, that, binaryPoint),
+      binaryPoint
+    )
+  }
+
+  override def <<(that: KnownWidth): IntervalRange = {
+    // TODO: (chick) requires being able to look at min/max of two possibly known bounds; see primop
```

Comment:
Unimplemented? Should implement properly?

Reply:
I implemented this.